import asyncio
from config import BOT_USERNAME, SUDO_USERS
from Zaid.decorators import authorized_users_only, sudo_users_only, errors
from Zaid.filters import command, other_filters
from Zaid.main import Test as USER
from pyrogram import filters
from pyrogram import Client
from pyrogram.errors import UserAlreadyParticipant
from pyrogram.types import *

@Client.on_message(filters.command("userbotjoin") & filters.group)
@authorized_users_only
async def join_chat(c: Client, m: Message):
    chat_id = m.chat.id
    try:
        invitelink = (await c.get_chat(chat_id)).invite_link
        if not invitelink:
            await c.export_chat_invite_link(chat_id)
            invitelink = (await c.get_chat(chat_id)).invite_link
        if invitelink.startswith("https://t.me/+"):
            invitelink = invitelink.replace(
                "https://t.me/+", "https://t.me/joinchat/"
            )
        await USER.join_chat(invitelink)
        return await c.send_message(chat_id, "‚úÖ userbot joined this chat")
    except UserAlreadyParticipant:
        return await c.send_message(chat_id, "‚úÖ userbot already in this chat")


@Client.on_message(command(["userbotleave",
                            f"leave@{BOT_USERNAME}"]) & filters.group & ~filters.edited)
@authorized_users_only
async def leave_one(client, message):
    try:
        await USER.send_message(message.chat.id, "‚úÖ userbot successfully left chat")
        await USER.leave_chat(message.chat.id)
    except BaseException:
        await message.reply_text(
            "‚ùå **userbot couldn't leave your group, may be floodwaits.**\n\n**¬ª or manually kick userbot from your group**"
        )

        return


@Client.on_message(command(["leaveall", f"leaveall@{BOT_USERNAME}"]))
@sudo_users_only
async def leave_all(client, message):
    if message.from_user.id not in SUDO_USERS:
        return

    left = 0
    failed = 0
    lol = await message.reply("üîÑ **userbot** leaving all chats !")
    async for dialog in USER.iter_dialogs():
        try:
            await USER.leave_chat(dialog.chat.id)
            left += 1
            await lol.edit(
                f"Userbot leaving all group...\n\nLeft: {left} chats.\nFailed: {failed} chats."
            )
        except BaseException:
            failed += 1
            await lol.edit(
                f"Userbot leaving...\n\nLeft: {left} chats.\nFailed: {failed} chats."
            )
        await asyncio.sleep(0.7)
    await client.send_message(
        message.chat.id, f"‚úÖ Left from: {left} chats.\n‚ùå Failed in: {failed} chats."
    )
